### PR TITLE
Add Q3_K_XS

### PR DESCRIPTION
**Pull request by ikawrakow**
TL;DR See #5055 

Before the recent two-bit quantization and importance matrix related changes, there were two low-bit quantization types available in `llama.cpp`: `Q2_K` and `Q3_K_S`. `Q2_K` was basically a 3-bit quantization with just the `attn_k` and `attn_q` tensors quantized with 2 bit. The table shows their model sizes and perplexities (`wiki.test.raw, n_ctx = 512`) for LLaMA-v2-70B:

| Quantization | Model size (GiB) | Perplexity |
| ---: | ---: | ---: | 
| Old Q2_K | 27.11 | 3.8164 |
| Old Q3_K_S | 27.70 | 3.7800 |

After the recent changes, `Q2_K` has become an actual 2-bit quantization (less than 3 bits-per-weight), has a LLaMA-v-70B model size of 23.71 GiB, and a perplexity of `4.0039` (using an importance matrix derived from `wiki.train.raw`). `Q3_K_S` has increased very slightly to 27.86 GiB, but has a better perplexity of `3.6603`. Based on #5005 there is a need to have an intermediate step in terms of model size between the new `Q2_K` and `Q3_K_S`. This PR adds such a quantization type as `Q3_K_XS`. The following table summarizes the new situation for LLaMA-v2-70B

| Quantization | Model size (GiB) | Perplexity |
| ---: | ---: | ---: | 
| Q2_K | 23.71 | 4.0039 |
| Q3_K_XS | 26.31 | 3.7808 |
| Q3_K_S | 27.86 | 3.6603 | 

The table on a graph:
q2_3_quants

### CODE REVIEW

**Comment by Nexesenex** [comment]
Screenshot 2024-01-09 at 13-58-22 r_LocalLLaMA - How much does Quantization actually impact models - KL Divergence Tests

Just a reminder of the table obtained after some optimizations you made on Q2_K and Q3_K_S a while ago.

**Comment by ikawrakow** [comment]
> Just a reminder of the table obtained after some optimizations you made on Q2_K and Q3_K_S in late August 2023. https://github.com/ggerganov/llama.cpp/pull/2807

I was taking the values from my notes, and I guess I forgot to update the notes when I made PR #2807. So, what we see in the above tables/graph is what we had before PR #2807. Here is an updated graph with the values post #2807 (i.e., current master)
q2_3_quants

**Review by ggerganov — APPROVED** [review]

**Comment by Artefact2** [comment]
Q3_K_XS seems to give broken results for mixtral-type models. Generation just ends immediately or prints a few symbols then stops.

I've tested and hit the bug with Noromaid-v0.4-Mixtral-Instruct-8x7b-Zloss, TenyxChat-8x7B and bagel-dpo-8x7b-v0.2.

**Comment by ikawrakow** [comment]
> Q3_K_XS seems to give broken results for mixtral-type models. Generation just ends immediately or prints a few symbols then stops.
> 
> I've tested and hit the bug with Noromaid-v0.4-Mixtral-Instruct-8x7b-Zloss, TenyxChat-8x7B and bagel-dpo-8x7b-v0.2.

Thank you for noticing. It should be fixed via PR #5113

**Comment by Artefact2** [comment]
I've tested the patch, it works. Thanks!
